src/api/c/cast.cpp

```diff
@@ -45,17 +47,51 @@ static af_array cast(const af_array in, const af_dtype type)
     }
 }
 
+static af_array castSparse(const af_array in, const af_dtype type)
+{
+    using namespace common;
+
+    const SparseArrayBase info = getSparseArrayBase(in);
+
+    if (info.getType() == type) {
+        return retain(in);
+    }
+
+    switch (type) {
+    case f32: return getHandle(castSparse<float  >(in));
+    case f64: return getHandle(castSparse<double >(in));
+    case c32: return getHandle(castSparse<cfloat >(in));
+    case c64: return getHandle(castSparse<cdouble>(in));
+    default: TYPE_ERROR(2, type);
+    }
+}
+
 af_err af_cast(af_array *out, const af_array in, const af_dtype type)
 {
     try {
-        const ArrayInfo info = getInfo(in);
+        const ArrayInfo info = getInfo(in, false, true);
+
+        af_dtype inType = info.getType();
+        if((inType == c32 || inType == c64)
```

Comment:
Using TYPE_ASSERT does not allow for verbose messages. The AF_ERROR used below is similar to TYPE_ASSERT (which itself is also a wrapper of AF_ERROR).